packages/hardhat-test-helpers/src/chai/matchers/bigNumber.ts

```diff
@@ -0,0 +1,264 @@
+
+  Assertion.overwriteMethod("closeTo", overrideCloseTo(utils));
+}
+
+type Methods = "eq" | "gt" | "lt" | "gte" | "lte";
+
+function throwIfUnsafe(i: number) {
+  if (typeof i === "number" && !Number.isSafeInteger(i)) {
+    throw new RangeError(
+      `Cannot compare to unsafe integer ${i}. Consider using BigInt(${i}) instead. For more details, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger`
+    );
+  }
+}
+
+function override(
+  method: Methods,
+  name: string,
+  negativeName: string,
+  utils: Chai.ChaiUtils
+) {
+  return (_super: (...args: any[]) => any) =>
+    overwriteBigNumberFunction(method, name, negativeName, _super, utils);
+}
+
+function normalize(
+  source:
+    | number
+    | bigint
+    | typeof BNType
+    | typeof EthersBigNumberType
+    | typeof BigNumberJsType
+    | string
+): bigint {
+  if (isEthersBigNumber(source)) {
+    return BigInt(source.toString());
+  } else if (isBN(source)) {
+    return BigInt(source.toString());
+  } else if (isBigNumberJsBigNumber(source)) {
+    return BigInt(source.toString(10));
+  } else if (
+    typeof source === "string" ||
+    typeof source === "number" ||
+    typeof source === "bigint"
+  ) {
+    return BigInt(source);
+  } else {
+    throw new Error(`cannot convert ${typeof source} to BigNumber`);
+  }
+}
+
+function isBigNumber(source: any): boolean {
+  return (
+    typeof source === "bigint" ||
+    isEthersBigNumber(source) ||
+    isBN(source) ||
+    isBigNumberJsBigNumber(source)
+  );
+}
+
+function overwriteBigNumberFunction(
+  functionName: Methods,
+  readableName: string,
+  readableNegativeName: string,
+  _super: (...args: any[]) => any,
+  chaiUtils: Chai.ChaiUtils
+) {
+  return function (this: Chai.AssertionStatic, ...args: any[]) {
+    const [actualArg] = args;
+    const expectedFlag = chaiUtils.flag(this, "object");
+    if (chaiUtils.flag(this, "doLength") && isEthersBigNumber(actualArg)) {
+      // TODO: consider whether we really need this case (support for eg
+      // `expect('foo').to.have.length.of.at.least(BigInt(2));`, and if so then
+      // get it tested and generalize it to work with all the different
+      // BigNumbers.
+      _super.apply(this, [actualArg.toNumber()]);
+      return;
+    }
+    function compare(method: Methods, lhs: bigint, rhs: bigint): boolean {
+      if (method === "eq") {
+        return lhs === rhs;
+      } else if (method === "gt") {
+        return lhs > rhs;
+      } else if (method === "lt") {
+        return lhs < rhs;
+      } else if (method === "gte") {
+        return lhs >= rhs;
+      } else if (method === "lte") {
+        return lhs <= rhs;
+      } else {
+        throw new Error(`Unknown comparison operation ${method}`);
+      }
+    }
+    if (isBigNumber(expectedFlag) || isBigNumber(actualArg)) {
+      const expected = normalize(expectedFlag);
+      const actual = normalize(actualArg);
```

Comment:
Isn't this inverted? If you do `expect(x).to.equal(y)`, `x` is the actual value and `y` is the expected one, but here they are swapped. This results in the wrong error message.

Reply:
I think it's not inverted. Note that the tests also exercise comparisons of `number` vs `number`, which induces a fallback to `_super.apply()`, and the tests are structured such that the error message is expected to be consistent across both `number`-vs-`number` and `BigNumber` comparisons.

Reply:
Hmm, I'm not convinced. If I do `expect(2).to.equal(3)` with chai without any extensions, this is the error:

![image](https://user-images.githubusercontent.com/417134/159003882-bae4d85b-1600-4760-a9c8-46028cd0f0d8.png)

But if I add `chai.use(bnChai)` and do `expect(2).to.equal(new BN(3))` I get:

![image](https://user-images.githubusercontent.com/417134/159003948-0ba9fd0a-dbed-43eb-a677-d8fdb1ba4323.png)

Notice that the actual and expected are inverted now.

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/2428/commits/5031927991113e6726e6a7be83b3263df1eeaf5b